pkg/cafs/cafs.go

```diff
@@ -4,89 +4,51 @@ import (
 	"context"
 	"fmt"
 	"io"
-	"log"
+	"sync"
 
 	lru "github.com/hashicorp/golang-lru"
 
 	"go.uber.org/zap"
 
+	"github.com/oneconcern/datamon/pkg/dlogger"
 	"github.com/oneconcern/datamon/pkg/storage"
 	"github.com/oneconcern/datamon/pkg/storage/localfs"
 
 	"github.com/docker/go-units"
 )
 
 const (
-	DefaultLeafSize    = 2 * 1024 * 1024
-	MaxLeafSize        = 5 * 1024 * 1024
+	// DeduplicationBlake is the deduplication scheme using the blake hash
+	// https://en.wikipedia.org/wiki/BLAKE_(hash_function).
+	//
+	// The implementation of the Blake hash we use (https://github.com/minio/blake2b-simd)
+	// is 3 to 5 times faster than usual hashes such as MD5 or SHA's.
 	DeduplicationBlake = "blake"
-)
-
-// LeafSize configuration for the blake2b hashes
-func LeafSize(sz uint32) Option {
-	return func(w *defaultFs) {
-		w.leafSize = sz
-	}
-}
-
-type HasOption func(*hasOpts)
-
-func HasOnlyRoots() HasOption {
-	return func(opts *hasOpts) {
-		opts.OnlyRoots = true
-	}
-}
-
-func HasGatherIncomplete() HasOption {
-	return func(opts *hasOpts) {
-		opts.OnlyRoots = true
-		opts.GatherIncomplete = true
-	}
-}
-
-func LeafTruncation(a bool) Option {
-	return func(w *defaultFs) {
-		w.leafTruncation = a
-	}
-}
-
-func Prefix(prefix string) Option {
-	return func(w *defaultFs) {
-		w.prefix = prefix
-	}
-}
-
-func Backend(store storage.Store) Option {
-	return func(w *defaultFs) {
-		w.store.backend = store
-	}
-}
 
-func ConcurrentFlushes(concurrentFlushes int) Option {
-	return func(w *defaultFs) {
-		w.concurrentFlushes = concurrentFlushes
-	}
-}
+	// DefaultCacheSize sets the default target LRU buffer cache in bytes.
+	//
+	// This defines the number of leaf buffers allocated to the cache (rounded up)
+	DefaultCacheSize = 50 * units.MiB
 
-func ReaderConcurrentChunkWrites(readerConcurrentChunkWrites int) Option {
-	return func(w *defaultFs) {
-		w.readerConcurrentChunkWrites = readerConcurrentChunkWrites
-	}
-}
+	// DefaultLeafSize sets the the default size of a blob leaf (2 MB). It cannot exceed MaxLeafSize.
+	// The actual leaf size used is usually specified by each bundle.
+	DefaultLeafSize uint32 = 2 * units.MiB
 
-type hasOpts struct {
-	OnlyRoots, GatherIncomplete bool
-	_                           struct{} // disallow unkeyed usage
-}
+	// MaxLeafSize is the maximum size of a buffer in the memory pool
+	MaxLeafSize = 5 * units.MiB
```

Comment:
We can set this to be the same as leaf size?

Reply:
@kerneltime this is used only by the FreeList. It corresponds to the maximum supported size to allocate leaf buffers. When the freelist kicks in, it dynamically adopts a buffer size not too much larger than the actual leaf size of the bundle.

Reply:
won't fix